compiler/src/parsing/lexer.re

```diff
@@ -214,6 +214,8 @@ let rec token = lexbuf => {
   | (unsigned_float, 'w') => positioned(WASMF32(sub_lexeme(lexbuf, 0, -1)))
   | (unsigned_float, 'W') => positioned(WASMF64(sub_lexeme(lexbuf, 0, -1)))
   | (unsigned_int, 't') => positioned(BIGINT(sub_lexeme(lexbuf, 0, -1)))
+  | (unsigned_int, '/', unsigned_int, 'r') =>
```

Comment:
This unfortunately doesn't support having a negative denominator. I know it sounds wild, but I think what we want to do here is specifically parse something like `5r` specifically as a rational denominator, and parse this the same way we handle `Number` rationals, but if you used a rational denominator you get a `Rational` literal instead of `Number` literal.

Reply:
Couldn't you just make the numerator negative though if you wanted a negative number?

Reply:
Also with what you're suggesting how should an error be thrown in the case of `let a = 5r`? Unless I am missing something, it would seem that every single token would have to scanned individually to make sure that a rational denominator has a corresponding numerator.

Reply:
> Couldn't you just make the numerator negative though if you wanted a negative number?

Of course, but we allow negative denominators for Number rational literals, so we should allow it here too.

Reply:
> Also with what you're suggesting how should an error be thrown in the case of `let a = 5r`? Unless I am missing something, it would seem that every single token would have to scanned individually to make sure that a rational denominator has a corresponding numerator.

In this case, we can provide a nice (well-formedness) error and say that this should be a fraction. You wouldn't really need to do anything extra in the parser.

It does though have the negative side effect of forcing us to add that to the parsetree.

An alternate proposal: OCaml just lexes numbers as numbers and accepts a range of suffixes, and carries the suffix with the token, to be interpreted by the typechecker. Maybe we should do that too—then what I'm suggesting makes far more sense. Ast_helper's `Expression.rational` then will accept an un-suffixed number or an `r`-suffixed number in the denominator, and the well-formedness check can see any number with an `r` suffix and can report that it should be a fraction.

Does that make sense?

Reply:
If I'm understanding what you're saying correctly then this would be a pretty big change to the parser, right? Would it have any real advantages over just having the well-formedness error other than not having to have a `PExpRationalDenom` variant? Also what about just adding another lexer rule like
```
| (unsigned_int, '/', '-', unsigned_int, 'r')
```
to allow for negative denominators?

Reply:
Not particularly. I also thought about what you mentioned here... what worries me is operator precedence. 

For `Number`, writing `4 * 2 / 3` parses as `(4 * 2) / 3` because multiplication and division share precedence, and rational literals share that same division precedence. But if you were to write `4 * 2 / 3r` suddenly this parses as `4 * (2 / 3r)` because there's a token that was grabbed by the lexer. Of course, that doesn't work out with the typing, and if it did you'd get the same result. But that's only case if `/` is defined as division, and things can get wonky. So we really want them to work the same. Should we just _also_ lex Number rationals as a single token? Maybe. I'd really need to think about it though. In the operator_precedence doc we do mention that exponentiation has higher precedence than division, though not implemented. You'd likely expect `2 / 3 ** 3` to be `2 / 27`, but if we lex 2/3 as its own token, it'd be `(2/3) ** 3` which is not `2/27`.

Reply:
Sorry, still a bit confused on some of the details you're proposing.
> lexes numbers as numbers and accepts a range of suffixes, and carries the suffix with the token, to be interpreted by the typechecker

I may be misunderstanding, but by this do you mean to say that the various `PConst...` for number types should be removed and the string representing the number should just contain the suffix in the token string like [in OCaml](https://github.com/ocaml/ocaml/blob/trunk/parsing/parsetree.mli#L26)? I'm not sure how this would help necessarily.

And then also to create a new well-formedness check that would reject instances of an `r`-suffixed token (since any valid uses of rational literals would have already been turned into rational number expressions)?

Reply:
That's what I was thinking, but now that I've thought about it more I don't think it's a great solution, or at least, it doesn't belong in this PR. I don't really like the solution of lexing a denominator token either.

I think it's fine to just ignore operator precedence for the sake of the Rational type, and you can lex it as a single token (just allowing negative numerators and/or denominators). If you wrote `2 * 3/4r` you definitely always mean `2 * (3/4r)` and never `(2 * 3)/4r`  syntax error.